### PR TITLE
Generic AS support for CPU devices

### PR DESCRIPTION
**Pull request by franz**

### CODE REVIEW

File: lib/CL/pocl_image_util.c [hunk 1]
```diff
@@ -255,6 +255,7 @@ pocl_get_image_information (cl_channel_order ch_order,
 /****************************************************/
 
 #define FOR4 unsigned i; for (i = 0; i < 4; i++)
+#define FOR2 unsigned i; for (i = 0; i < 2; i++)
```
Comment (pjaaskel) [inline lib/CL/pocl_image_util.c:258]: What is this macro trickery for? Pls add a comment above.

**Review by pjaaskel — APPROVED** [review]

**Comment by pjaaskel** [comment]
Clang 14 fails an assertion with this branch when building the bc lib (Release build mode, assertions ON):

```
[ 18%] Building CL to LLVM bitcode /home/visit0r/src/pocl/test-pr-cmake-llvm-14-tce-tests-build/lib/kernel/host/broadwell/bitselect.cl.bc
cd /home/visit0r/src/pocl/test-pr-cmake-llvm-14-tce-tests-build/lib/kernel/host && /home/visit0r/local/stow/llvmtce-14/bin/clang --target=x86_64-unknown-linux-gnu -D_CL_DISABLE_HALF -march=broadwell -emit-llvm -ffp-contract=off -Wall -Wno-unused-local-typedef -Xclang -cl-std=CL3.0 -D__OPENCL_C_VERSION__=300 -DENABLE_SLEEF=1 -xcl -fno-stack-protector -fPIC -Wno-psabi -D__OPENCL_VERSION__=300 -DPOCL_DEVICE_ADDRESS_BITS=64 -Dcl_khr_int64=1 -Dcl_khr_byte_addressable_store=1 -Dcl_khr_global_int32_base_atomics=1 -Dcl_khr_global_int32_extended_atomics=1 -Dcl_khr_local_int32_base_atomics=1 -Dcl_khr_local_int32_extended_atomics=1 -Dcl_khr_3d_image_writes=1 -Dcl_khr_command_buffer=1 -Dcl_khr_spir=1 -Dcl_khr_int64_base_atomics=1 -Dcl_khr_int64_extended_atomics=1 -Dcl_khr_fp64=1 -D__opencl_c_3d_image_writes=1 -D__opencl_c_images=1 -D__opencl_c_atomic_order_acq_rel=1 -D__opencl_c_atomic_order_seq_cst=1 -D__opencl_c_atomic_scope_device=1 -D__opencl_c_program_scope_global_variables=1 -D__opencl_c_generic_address_space=1 -D__opencl_c_read_write_images=1 -D__opencl_c_int64=1 -D__opencl_c_fp64=1 -Xclang -cl-ext=-all,+cl_khr_byte_addressable_store,+cl_khr_global_int32_base_atomics,+cl_khr_global_int32_extended_atomics,+cl_khr_local_int32_base_atomics,+cl_khr_local_int32_extended_atomics,+cl_khr_3d_image_writes,+cl_khr_command_buffer,+cl_khr_spir,+cl_khr_int64_base_atomics,+cl_khr_int64_extended_atomics,+cl_khr_fp64,+__opencl_c_3d_image_writes,+__opencl_c_images,+__opencl_c_atomic_order_acq_rel,+__opencl_c_atomic_order_seq_cst,+__opencl_c_atomic_scope_device,+__opencl_c_program_scope_global_variables,+__opencl_c_generic_address_space,+__opencl_c_read_write_images,+__opencl_c_int64,+__opencl_c_fp64, -o /home/visit0r/src/pocl/test-pr-cmake-llvm-14-tce-tests-build/lib/kernel/host/broadwell/bitselect.cl.bc -c /home/visit0r/src/pocl/lib/kernel/bitselect.cl -include /home/visit0r/src/pocl/include/_kernel.h -include /home/visit0r/src/pocl/include/_enable_all_exts.h
clang: /home/visit0r/src/tce/tce/llvm-build-Release/release_14/llvm/lib/IR/Instructions.cpp:3154: static llvm::CastInst* llvm::CastInst::Create(llvm::Instruction::CastOps, llvm::Value*, llvm::Type*, const llvm::Twine&, llvm::Instruction*): Assertion `castIsValid(op, S, Ty) && "Invalid cast!"' failed.
...
```
With an asserts OFF LLVM it goes forward. If it's a difficult fix, just disable the generic AS feature for older LLVMs.

File: lib/CL/pocl_image_util.c (hunk 1 above)
Comment (franz) [reply lib/CL/pocl_image_util.c:258]: AFAICT nothing except save some typing. I can remove it if you think.

Comment (pjaaskel) [reply lib/CL/pocl_image_util.c:258]: Yep seems strange and might make the code reader to ask the same thing I asked: Is there some additional macro magic hidden somewhere?

**Comment by franz** [comment]
Interestingly the crashes were only with LLVM version 14, older (back to 10) don't crash. The code that was tripping it up is `*(itype*)&vtype`, a bitcast of vector done via taking address + dereferencing. Replacing it with `as_type` solved the crash. There's still one test failing however, i'm looking into that.

**Pull request merged**